### PR TITLE
Provide yes man endpoint for inflexible load balancers

### LINKED ISSUE #840: Need health check endpoint not protected by auth
When I tried the HEAD version with http basic auth support, I noticed that everything is protected by auth. When I want to deploy the Docker registry behind a load balanced like ELB, however, I need a health check URL that works without auth.

Could you please implement one before you release the http basic auth feature?

### PR DESCRIPTION
**Pull request by stevvooe**
Certain load balancers, such as Amazon's Elastic Load Balancer, have a very
limited notion of health. While a properly configured and operational registry
should always return a 401 when hitting "/v2/", such load balancers cannot be
configured to treat this response code as healthy. This changeset makes "/"
always return a 200 response, unless the health checks have failed.

Closes #840 

Signed-off-by: Stephen J Day <stephen.day@docker.com>

### CODE REVIEW

**Comment by RichardScothern** [comment]
I'm happy with the code and intent.

I think we should document this somewhere, maybe a footnote in deploying.md?

**Comment by dmp42** [comment]
I would love to have "recipes":

 * yes man check endpoint
 * no-auth pull with auth push setup
 * load balancing and scaling tips and tricks
 * read-only registry
 * proxy cache registry (?)

**Comment by stevvooe** [comment]
@RichardScothern Do you want me to add this as part of this PR?

**Comment by RichardScothern** [comment]
The recipes aren't required for this PR, but a note somewhere for ELB users to know about this feature would be nice.

**Comment by stevvooe** [comment]
@RichardScothern I'm not sure we talk about load balancing in "deploying"? Is there another place where this might belong?

**Comment by RichardScothern** [comment]
I was thinking a monitoring subsection under  https://github.com/docker/distribution/blob/master/docs/deploying.md#running-a-domain-registry

Something like ""to healthcheck a properly configured registry, monitor the authorized /v2 endpoint.  If your load balancer only accepts 200 as a valid healthcheck / is ELB you can point it to /""

**Comment by stevvooe** [comment]
@RichardScothern Added a guide.

File: docs/deploying.md [hunk 1]
```diff
@@ -84,6 +84,50 @@ A certificate issuer may supply you with an *intermediate* certificate. In this
...
+  - Storage Driver
+  - HTTP Secret
+  - Redis Cache (if configured)
+
+If any of these are different, the registry may have trouble serving requests.
+As an example, if you're using the filesystem driver, all registry instances
+must have access to the same filesystem root, which means they should be in
+the same machine. For other drivers, such as s3 or azure, they should be
+accessing the same resource, and will likely share an identical configuration.
+The _HTTP Secret_ coordinates downloads, so also must be the same across
```
Comment (RichardScothern) [inline docs/deploying.md]: Doesn't this coordinate uploads?

File: docs/deploying.md [hunk 2]
```diff
@@ -84,6 +84,50 @@ A certificate issuer may supply you with an *intermediate* certificate. In this
...
+instances. Configuring different redis instances will mostly work (at the of
```
Comment (RichardScothern) [inline docs/deploying.md]: 'time of'

File: docs/deploying.md [hunk 3]
```diff
@@ -84,6 +84,50 @@ A certificate issuer may supply you with an *intermediate* certificate. In this
...
+Getting the headers correct is very important. For all responses to any
+request under the "/v2/" url space, the `Docker-Distribution-API-Version`
+header should be set to the value "registry/2.0", even for a 4xx response.
+This header allows the docker engine to quickly resolve authentication realms
+and fallback to version 1 registries, if necessary. Confirming this is setup
+correctly can help avoid problems with fallback.
+
+A properly secured registry should return 401 when the "/v2/" endpoint is hit
+without credentials. The response should include a `WWW-Authenticate`
+challenge, providing quidance on how to authentication, such as with basic
```
Comment (RichardScothern) [inline docs/deploying.md]: s/quidance/guidance
s/authentication/authenticate

**Comment by RichardScothern** [comment]
Couple of nits, but LGTM.
ping @dmp42 @aaronlehmann

File: docs/deploying.md (hunk 1 above)
Comment (stevvooe) [inline docs/deploying.md]: Yes.

**Comment by dmp42** [comment]
Guide LGTM

**Pull request merged**